### ISSUE TITLE
Slow response times for Admin UI pages

### ISSUE THREAD

**Issue by ttsirkia** [keystonejs/keystone#1197]
I noticed that even with an empty database it takes about 700 ms before browser gets response from the Admin UI. I've tested this with a few virtual machines and the results are the same. Keystone is running in production mode and the other views (my own) are served in less than 30 ms.

Any ideas what makes those pages so slow?

**Comment by morenoh149** [comment]
could you share a screencap of your network waterfall in dev tools. Both FF and chrome have this.

**Comment by ttsirkia** [comment]
There are three users in my database.

`GET /keystone/users 200 1228.269 ms`

![users](https://cloud.githubusercontent.com/assets/2797729/6613393/ca61ec58-c894-11e4-8d91-bb89cebb0628.png)

At the same time, serving my own index view takes about 30 ms.

**Comment by morenoh149** [comment]
😦 somethings wrong. What api endpoint (route) is `GET users` hitting? Could you also share the output of running the query in the mongo shell. http://docs.mongodb.org/manual/tutorial/analyze-query-plan/ aka `explain`. You're probably missing an index somewhere.

**Comment by ttsirkia** [comment]
`users` is something that the Admin UI generates out of the box. It has not my own code. As there are only three users (and actually nothing else), indices cannot be the problem.

I just used `yo` to create an empty project and I'm running it on a virtualized server. The database is empty: `GET /keystone/users 200 757.409 ms` This is a real virtualized server running Ubuntu 14.04. In the screenshot, there is a similar server but it runs my own laptop inside Virtualbox.

After getting the response from the server, it still takes many seconds before the admin view is fully rendered and JavaScript is ready.

So the response time from server is 20 - 40 times slower than any of my own views and the complete render time much worse.

**Comment by ttsirkia** [comment]
Can anyone else confirm this or is there just a common factor in my two environments which causes this?

**Comment by tivac** [comment]
I'm using io.js 1.5.1 with MongoDB 2.6.8 and seeing about a 2.5s wait before I get `/users` HTML back. I have 136 users in my DB.

**Comment by sreyemnayr** [comment]
I'm having the same issue.  Even on a local server, the navigation will appear fairly quickly, then like 6 seconds later the edit form finally shows up.

**Comment by lojack** [comment]
Same issue here, also getting a warning on the backend

```
Note: The code generator has deoptimised the styling of "{redacted}/node_modules/moment/moment.js" as it exceeds the max of "100KB".
```

On the list view, everything except for the Create button loads, that error displays and then shortly after the Create button shows up. The detail view takes a bit more time to load. 

Without digging deeper it seems as though its rebuilding the Javascript files on every single request.

![screen shot 2015-04-28 at 3 27 35 pm](https://cloud.githubusercontent.com/assets/180470/7378430/49d28fd8-edbb-11e4-8b0f-8014e6121491.png)
![screen shot 2015-04-28 at 3 28 05 pm](https://cloud.githubusercontent.com/assets/180470/7378429/49d12530-edbb-11e4-81ed-016f8720664c.png)

**Comment by GlennGeenen** [comment]
Every time I get a slow response I see the same error as @lojack.

Seems it has something to do with Babel: http://stackoverflow.com/questions/29576341/what-does-the-code-generator-has-deoptimised-the-styling-of-some-file-as-it-e

**Comment by ppiekarczyk** [comment]
`compact: false` will do the trick for you, but this needs to be PR'd and added into keystone.

**Comment by GlennGeenen** [comment]
Setting compact: false in admin/app/static.js fixed it for me. I redid my commits because of not having .editorconfig working.

**Comment by JedWatson** [comment]
I've rewritten our browserify build process (basically custom-rolled us a version of browserify-middleware that works _exactly_ the way I want it to) which should resolve this.

The new functionality 
* initialises the build when Keystone starts
* waits for builds to complete before serving files
* will be compatible with BYO Express setups
* supports watchify for rebuilds in dev mode, but not `NODE_ENV=development`
* allows for Streams in place of real files (unlocks building dynamic field bundles, which will let us exclude unused field types for improved performance)
* reports changes and rebuilds to the console in dev mode

The changes should speed up build and delivery considerably. I'll also take a look at that compact option before closing this.

If anyone can try out the new changes in `master` and confirm whether it improves things, that would be great.

**Comment by JedWatson** [comment]
Also re: inclusion of lots of client-side files, they're around for the filters which are still using jQuery, @jossmac and I are working on implementing Elemental UI and completing the transition to React which will let us rip out most of the code that's still bloating the initial download.

**Comment by GlennGeenen** [comment]
I no longer get the 'code generator has deoptimised' warning and the response times have improved greatly. Thanks a lot!!!

**Comment by JedWatson** [comment]
@GlennGeenen excellent :grinning:

**Comment by ttsirkia** [comment]
I tried the version from `master`. Loading `users` page takes now about 600 ms with one user. The process is still very slow:

![aika](https://cloud.githubusercontent.com/assets/2797729/7665694/add142d0-fbcc-11e4-92f0-170d309cea8c.png)

**Comment by ttsirkia** [comment]
Hmm, it seems that Firefox is always requesting all the JS files again (as in above, response code 200). For some reason it doesn't use cache although ETag is present.

In Chrome, after files are cached, the loading process is faster but still taking about 600 ms.

**Comment by tmvanetten** [comment]
I am not seeing this issue in 0.3.16 or 0.4.0.  Is this issue still relevant?

<img width="1436" alt="screen shot 2016-02-17 at 9 09 58 pm" src="https://cloud.githubusercontent.com/assets/3821200/13132726/ec9ed23e-d5ba-11e5-80b5-09cf0b8c2816.png">

**Comment by JedWatson** [comment]
I think we're good now :) thanks for the bump @tmvanetten!

**Comment by ttsirkia** [comment]
I don't think this is solved. Here is an example with 0.3.16. The requested model does not contain any rows. Same behavior with Firefox and Chrome.

![keystone](https://cloud.githubusercontent.com/assets/2797729/13135536/1cfe5a80-d61a-11e5-80a0-ba5fe02f50a8.png)

**Comment by tmvanetten** [comment]
@ttsirkia my bad your correct I am still seeing the issue on 0.3.16 as well.

<img width="1440" alt="screen shot 2016-02-19 at 11 16 27 am" src="https://cloud.githubusercontent.com/assets/3821200/13183079/aca6b55c-d6fa-11e5-8d00-bebd34733889.png">

**Comment by JedWatson** [comment]
TBH I don't think we'll be revisiting this for 0.3, 0.4 has had a _huge_ amount of work done on it and is very close to shipping.

**Comment by ttsirkia** [comment]
I think that is reasonable for 0.3 because the bug affects just the admin UI pages. I haven't tried what is the situation with 0.4.

**Comment by ttsirkia** [comment]
@JedWatson I just tried this with the test project for 0.4 and there everything works just fine. I'm OK if this issue is closed with won't fix status as the new version is not affected.

**Comment by JedWatson** [comment]
Thanks @ttsirkia. Not sure "wont fix" gives the right impression - it's fixed, just not back ported for 0.3... I care a lot about the performance of Keystone's Admin UI, am just focused on pushing forward with the new version :)

**Comment by ttsirkia** [comment]
That's perfectly fine. It is true that it is fixed. Won't fix for 0.3, fixed in 0.4. :-)

**Comment by wub** [comment]
Stumbled upon this - having issues with the latest version (as of Jun 28 '16) - but *only* on OS X (macOS...).

When I run it on Windows, I get the 50ms response times people are seeing. OS X? 700ms, 800ms+
